Support GCS

raghunandanbhat opened the pull request:
fixes #38

raghunandanbhat commented:
@zhousun please review when you get a chance. thanks!

zhousun commented:
@raghunandanbhat thanks for the contribute. For delta-rs, I think it is possible to inline the secret as json instead of providing a path: https://docs.rs/object_store/latest/src/object_store/gcp/builder.rs.html#365

raghunandanbhat commented:
delta-rs now uses the service account json instead of the file path.

I believe it is better to provide the path when creating/inserting a secret into `mooncake.secrets`, as it is easier than typing the entire JSON.

zhousun commented:
If I understand correct, providing a path like this requires the files to be in server side. For example if you connect to a remote postgres, you cannot use a secret file in local path. That's why I believe having a way to direct type/copy the secret is important.

zhousun reviewed (commented):
In general lgtm, will approve once you resolve the other comment (service account json file won't work with server-client setup).
thx!

raghunandanbhat commented:
> If I understand correct, providing a path like this requires the files to be in server side. For example if you connect to a remote postgres, you cannot use a secret file in local path. That's why I believe having a way to direct type/copy the secret is important.

what if we keep both options? 
- type/copy secrets for server-client setup
- path based secrets for local setup

raghunandanbhat commented on README.md:
````diff
@@ -79,10 +79,21 @@ SELECT * from user_activity;
 
 > **Note**: On Neon, only cloud storage is supported. Neon users must bring their own S3 or R2 buckets or get a free S3 bucket by signing up at [s3.pgmooncake.com](https://s3.pgmooncake.com/). For cloud storage configuration instructions, see [Cloud Storage](https://pgmooncake.com/docs/cloud-storage). We are working to improve this experience.
 
+#### GCS Buckets
+- Add your GCS credentials
+```sql
+SELECT mooncake.create_secret('<name>', 'GCS', '<HMAC_key_id>', '<HMAC_secret>', '{"PATH":"path/to/service_account.json"}');
+```
+- Set default bucket after adding GCS credentials.
+```sql
+SET mooncake.default_bucket = 'gs://<gcs_bucket>';
+```
+>**Note**: delta-rs seems to accept only Service Account JSON credentials when creating delta tables in GCS.
+
````
sure. 
please make sure to update the sql query.

```sql
SELECT mooncake.create_secret('<name>', 'GCS', '<HMAC_key_id>', '<HMAC_secret>', '{"PATH":"path/to/service_account.json"}');
```
to
```sql
SELECT mooncake.create_secret('<name>', 'GCS', '<HMAC_key_id>', '<HMAC_secret>', '{"GCS_SECRET":"Service Account JSON"}');
```

zhousun replied:
cc @paurora17

zhousun commented:
> > If I understand correct, providing a path like this requires the files to be in server side. For example if you connect to a remote postgres, you cannot use a secret file in local path. That's why I believe having a way to direct type/copy the secret is important.
> 
> what if we keep both options?
> 
> * type/copy secrets for server-client setup
> * path based secrets for local setup

Yep that would work

dpxcc commented on rust_extensions/delta/Cargo.toml:
```diff
@@ -8,7 +8,7 @@ crate-type = ["staticlib"]
 [dependencies]
 cxx = "1.0"
-deltalake = {version = "0.21", features = ["s3"] }
+deltalake = {version = "0.21", features = ["s3", "gcs"] }
```
nit: could we sort the features alphabetically?

dpxcc commented on rust_extensions/delta/src/lib.rs:
```diff
@@ -33,7 +34,8 @@ mod ffi {
     // Register S3 handlers
-    register_handlers(None);
+    register_aws_handlers(None);
+    register_gcp_handlers(None);
```
nit: need to update the comment, or better just remove the comment since the code is quite clear

dpxcc commented on rust_extensions/delta/src/lib.rs:
```diff
@@ -48,8 +50,10 @@ pub fn DeltaCreateTable(
             serde_json::from_str(options.to_str()?).expect("invalid options");
-        // Write directly to S3 without locking is safe since Mooncake is the only writer
-        storage_options.insert("AWS_S3_ALLOW_UNSAFE_RENAME".to_string(), "true".to_string());
+        if !path.is_empty() && path.to_string().starts_with("s3://") {
```
nit: can we extract
```rust
!path.is_empty() && path.to_string().starts_with("s3://")
```
to a helper function?

raghunandanbhat replied:
done

dpxcc commented on sql/pg_mooncake--0.1.0.sql:
```diff
@@ -319,9 +321,32 @@ BEGIN
+            name,
+            type,
+            coalesce(extra_params->>'SCOPE', ''),
+            format('CREATE SECRET "duckdb_secret_%s" (TYPE %s, KEY_ID %L, SECRET %L', name, type, key_id, secret)||
```
nit: space before `||`

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -282,17 +284,24 @@ string ColumnstoreMetadata::SecretsSearchDeltaOptions(const string &path) {
             string scope = TextDatumGetCString(values[2]);
-            if ((scope.empty() || StringUtil::StartsWith(path, scope)) && longest_match <= scope.length()) {
+            if ((scope.empty() || StringUtil::StartsWith(path, scope)) && longest_match_s3 <= scope.length()) {
+                option = TextDatumGetCString(values[4]);
+                longest_match_s3 = scope.length();
+            }
+        } else if (strcmp(TextDatumGetCString(values[1]), "GCS") == 0) {
+            string scope = TextDatumGetCString(values[2]);
+            if ((scope.empty() || StringUtil::StartsWith(path, scope)) && longest_match_gcs <= scope.length()) {
                 option = TextDatumGetCString(values[4]);
-                longest_match = scope.length();
+                longest_match_gcs = scope.length();
```
there is no need to treat s3 and gcs differently here
1. I suppose their scopes are disjoint
2. Even if their scopes are not disjoint, your current code is not necessarily picking the longest overall

raghunandanbhat replied:
fixed

raghunandanbhat commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -263,6 +263,8 @@ vector<string> ColumnstoreMetadata::SecretsGetDuckdbQueries() {
         heap_deform_tuple(tuple, desc, values, isnull);
         if (strcmp(TextDatumGetCString(values[1]), "S3") == 0) {
             queries.emplace_back(TextDatumGetCString(values[3]));
+        } else if (strcmp(TextDatumGetCString(values[1]), "GCS") == 0) {
```
done

dpxcc commented on rust_extensions/delta/src/lib.rs:
```diff
@@ -48,8 +55,10 @@ pub fn DeltaCreateTable(
             serde_json::from_str(options.to_str()?).expect("invalid options");
-        // Write directly to S3 without locking is safe since Mooncake is the only writer
-        storage_options.insert("AWS_S3_ALLOW_UNSAFE_RENAME".to_string(), "true".to_string());
+        if get_storage_type(path) == StorageType::S3 {
```
Thinking about this again, I think it might be better to simply return the type from `ColumnstoreMetadata::SecretsSearchDeltaOptions()` and pass it down here

raghunandanbhat replied:
yes, it is much simpler. we can return `tuple<string, string>` from `ColumnstoreMetadata::SecretsSearchDeltaOptions()` and pass them separately to `DeltaCreateTable()`

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -288,12 +289,15 @@ string ColumnstoreMetadata::SecretsSearchDeltaOptions(const string &path) {
-                longest_match = scope.length();
-            }
+        const char *bucket_type = TextDatumGetCString(values[1]);
+        if ((strcmp(bucket_type, "S3") == 0 && !StringUtil::StartsWith(path, "s3://")) ||
+            (strcmp(bucket_type, "GCS") == 0 && !StringUtil::StartsWith(path, "gs://"))) {
```
This is inconsistent with `get_storage_type()` where there are other prefixes

raghunandanbhat commented on rust_extensions/delta/src/lib.rs:
```diff
@@ -165,3 +176,16 @@ fn convert_postgres_to_delta_type(column_type: &str) -> DataType {
+
+    let path_str = path.to_string();
+    match path_str {
+        s if s.starts_with("s3://") || s.starts_with("s3a://") => StorageType::S3,
```
I thought `s3n://` is older and no longer being used. If duckdb supports it, we should too. will include this

dpxcc commented on README.md:
```diff
@@ -79,10 +79,11 @@ SELECT * from user_activity;
 
 > **Note**: On Neon, only cloud storage is supported. Neon users must bring their own S3 or R2 buckets or get a free S3 bucket by signing up at [s3.pgmooncake.com](https://s3.pgmooncake.com/). For cloud storage configuration instructions, see [Cloud Storage](https://pgmooncake.com/docs/cloud-storage). We are working to improve this experience.
 
+
```
nit: rm extra newline

dpxcc commented on rust_extensions/delta/src/lib.rs:
```diff
@@ -48,8 +52,10 @@ pub fn DeltaCreateTable(
             serde_json::from_str(options.to_str()?).expect("invalid options");
-        // Write directly to S3 without locking is safe since Mooncake is the only writer
-        storage_options.insert("AWS_S3_ALLOW_UNSAFE_RENAME".to_string(), "true".to_string());
+        if storage_type.to_str()?.eq("S3") {
```
I suppose you can just `storage_type.eq("S3")`

dpxcc commented on sql/pg_mooncake--0.1.0.sql:
```diff
@@ -275,14 +275,17 @@ RETURNS VOID
 DECLARE
-    allowed_keys TEXT[] := ARRAY['ENDPOINT', 'REGION', 'SCOPE', 'USE_SSL'];
+    s3_allowed_keys TEXT[] := ARRAY['ENDPOINT', 'REGION', 'SCOPE', 'USE_SSL'];
+    gcs_allowed_keys TEXT[] := ARRAY['SCOPE', 'GCS_SECRET', 'PATH'];
```
nit: sort alphabetically

dpxcc commented on sql/pg_mooncake--0.1.0.sql:
```diff
@@ -275,14 +275,17 @@ RETURNS VOID
     invalid_keys TEXT[];
     delta_endpoint TEXT;
+    gcs_service_account_json JSONB;
+    gcs_required_keys TEXT[] := ARRAY['type', 'project_id', 'private_key_id', 'private_key', 'client_email', 'client_id'];
```
Sort alphabetically
Should these also be uppercase?
We need to also add doc for these - I think your original doc doesn't include them

raghunandanbhat replied:
`gcs_required_keys` lists the keys in lowercase and in the same order as they appear in secrets JSON file. This makes it easy to spot which key is missing when error message is shown. 

cc: @paurora17 
For docs, you can use the below- 

#### Google Cloud Storage
Add your GCS credentials by either providing the path to the service account JSON file:
```sql
SELECT mooncake.create_secret('<name>', 'GCS', '<HMAC_key_id>', '<HMAC_secret>', '{"PATH":"path/to/service_account.json"}');
```
Or by supplying the Service Account JSON string directly:
```sql
SELECT mooncake.create_secret('<name>', 'GCS', '<HMAC_key_id>', '<HMAC_secret>', '{"GCS_SECRET":"Service Account JSON string"}');
```
>Note: When supplying secrets using a JSON string, ensure it includes the following keys, properly escaped: `'type', 'project_id', 'private_key_id', 'private_key', 'client_email', 'client_id'`.

Set the default bucket after adding GCS credentials:
```sql
SET mooncake.default_bucket = 'gs://<gcs_bucket>';
```
>**Note**: delta-rs seems to accept only Service Account JSON credentials when creating delta tables in GCS.

dpxcc replied:
Thanks so much for the background! I'll revert the order of keys to your original order

@paurora17 
When adding the docs, please clearly state that `'{"PATH":"path/to/service_account.json"}'` only works if `service_account.json` is located on the server side
By default, this won't work in Docker containers unless the file is mounted, and it definitely won't work on Neon
Should we consider omitting this option from the doc altogether?

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -271,35 +286,39 @@ vector<string> ColumnstoreMetadata::SecretsGetDuckdbQueries() {
-                longest_match = scope.length();
-            }
+        const string storage_type = TextDatumGetCString(values[1]);
+        if (!IsMatchingStorageType(storage_type, path)) {
```
You should just compute the type based on `path` outside of the loop

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -57,6 +57,20 @@ Oid Secrets() {
     return get_relname_relid("secrets", Mooncake());
 }
 
+static const std::vector<string> s3_prefixes = {"s3://", "s3a://", "s3n://"};
```
nit:
1. no need for static. you are already in anonymous namespace
2. std::vector -> vector

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -57,6 +57,20 @@ Oid Secrets() {
+static const std::vector<string> gcs_prefixes = {"gcs://", "gs://"};
+
+// Checks if a path matches the URL scheme for the given storage type (S3 or GCS)
+bool IsMatchingStorageType(const string &storage_type, const string &path) {
```
your variable names are not conforming with our code convention

dpxcc commented on src/columnstore/columnstore_metadata.cpp:
```diff
@@ -57,6 +57,20 @@ Oid Secrets() {
+bool IsMatchingStorageType(const string &storage_type, const string &path) {
+    auto has_matching_prefix = [&path](const std::vector<string> &prefixes) {
+        return std::any_of(prefixes.begin(), prefixes.end(),
+                           [&path](string prefix) { return StringUtil::StartsWith(path, prefix); });
```
should use reference `string &prefix`

dpxcc commented on src/columnstore/columnstore_metadata.hpp:
```diff
@@ -29,7 +29,7 @@ class ColumnstoreMetadata {
 
     vector<string> SecretsGetDuckdbQueries();
-    string SecretsSearchDeltaOptions(const string &path);
+    std::tuple<string, string> SecretsSearchDeltaStorageConfig(const string &path);
```
nit: `std::tuple<string /*type*/, string /*options*/>`

dpxcc commented on src/lake/lake.cpp:
```diff
@@ -19,14 +19,16 @@ class LakeWriter {
             auto [path, timeline_id] = metadata.TablesSearch(oid);
-            cached_table_infos[oid] = {path, std::move(timeline_id), metadata.SecretsSearchDeltaOptions(path)};
+            auto [storage_type, options] = metadata.SecretsSearchDeltaStorageConfig(path);
+            cached_table_infos[oid] = {path, std::move(timeline_id), storage_type, options};
```
should use `std::move()` on `storage_type` and `options`

dpxcc commented:
Sorry, but my force-push to gcs-support branch somehow triggered Github to close this PR accidentally..

We've cleaned up your implementation at
https://github.com/Mooncake-Labs/pg_mooncake/commit/42392b796745c0304ee55c8e25faa24d1e27a75e
Let us know if you are okay with that

Also, I believe DuckDB's delta_scan() won't work with GCS right now due to https://github.com/duckdb/duckdb-delta/issues/97

raghunandanbhat commented:
>We've cleaned up your implementation at https://github.com/Mooncake-Labs/pg_mooncake/commit/42392b796745c0304ee55c8e25faa24d1e27a75e
(The cleanup addressed the comments above)
Let us know if you are okay with that

looks great!

dpxcc commented:
Merged into the main branch. Thank you for your work! 🥮

malpr commented:
Hi, could you clarify why the GCS support was reverted?
